Optimize ps.array.intersect

cwarden opened the pull request:
Use an intersection algorithm optimized for sorted arrays of primitives.

I think this change is safe.  Is it possible for the arguments to intersect, called from ps.Cube.prototype._getPos to be unsorted?

cwarden committed:
Optimize ps.array.intersect

Use an intersection algorithm optimized for sorted arrays of primitives.

thesmart commented:
Hey Christian, thanks much for the patch.  Makes sense.

Would highly recommend considering .pop versus .shift:
http://jsperf.com/pop-vs-shift-on-a-array/11

On Wed, Jul 18, 2012 at 4:14 PM, Christian G. Warden <
reply@reply.github.com

> wrote:
> 
> Use an intersection algorithm optimized for sorted arrays of primitives.
> 
> I think this change is safe.  Is it possible for the arguments to
> intersect, called from ps.Cube.prototype._getPos to be unsorted?
> 
> You can merge this Pull Request by running:
> 
>   git pull https://github.com/cwarden/js-hypercube intersect-optimization
> 
> Or you can view, comment on it, or merge it online at:
> 
>   https://github.com/thesmart/js-hypercube/pull/2
> 
> -- Commit Summary --
> - Optimize ps.array.intersect
> 
> -- File Changes --
> 
> M src/ps.array.js (15)
> 
> -- Patch Links --
> 
>   https://github.com/thesmart/js-hypercube/pull/2.patch
>   https://github.com/thesmart/js-hypercube/pull/2.diff
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/thesmart/js-hypercube/pull/2

cwarden committed:
Use .pop instead of .shift in intersect

Use Array.prototype.pop instead of Array.prototype.shift when finding
intersection of sorted arrays because it is much faster.  A benchmark of
the two methods in addition to the original algorithm is at
http://bit.ly/P9FrZK.

cwarden commented:
@thesmart, good suggestion on pop vs shift.

cwarden committed:
Keep result of intersect sorted
